docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,215 @@
+- Worker: Machine-Controller-Manager related objects: machineclasses, machinedeployments, machinesets, machines
+
+This list depends on the currently installed extensions and can change in the future
+
+## Proposal
+
+### CRD on the garden cluster
+
+The Garden cluster has a new Custom Resource which is stored in the project namespace of the Shoot. It contains all the required data described above so that the the control plane can be recreated on the New Seed.
+The metadata is separated into two sections. The first is generated by the Gardener Controller Manager and then either used to generate new resources (e.g secrets) or data in it is directly deployed to the Shoot's control plane on the New Seed.
+
+The second is generated by the extension controllers in the seed and then synced to the shoot by a new ShootState controller (described below).
+
+```
+apiVersion: core.gardener.cloud/v1alpha1
+kind: ShootState
+metadata:
+  name: my-shoot
+  namespace: garden-core
+gardener:
+- name: ca
+  data:
+    ca.crt: ...
+    ca.key: ...
+- name: ssh-keypair
+  data:
+    id_rsa: ...
+- name:
+...
+extensions:
+- kind: Infrastructure
+  state: ... (Terraform state)
+- kind: ControlPlane
+  purpose: normal
+  state: ... (Certificates generated by the extension)
+- kind: Worker
+  state: ... (Machine objects)
+```
+
+There can be sensitive data in the `ShootState` which has to be hidden from the end-users. Currently we plan to keep the `ShootState` resource encrypted in etcd.
+
+Another option is to have references to secrets in the `ShootState`. However, in this case secrets will have to be stored in dedicated namespaces, different than the project namespace of the `Shoot`.
+
+### Gardener extensions changes
+
+All extension controllers which require state migration must save their state in a new `status.state` field and act on an annotation `core.gardener.cloud/operation=restore` in the respective Custom Resources which should trigger a restoration operation instead of reconciliation. A restoration operation means that the extension has to restore its state in the Shoot's namespace on the __New Seed__ from the `status.state` field.
+
+**Note**: some states can be too large so it makes sense to consider adding them as configmap or secret references.
+
+As an example: the `Infrastructure` resource must save the terraform state.
+
+```
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Infrastructure
+metadata:
+  name: infrastructure
+  namespace: shoot--foo--bar
+spec:
+  type: azure
+  region: eu-west-1
+  secretRef:
+    name: cloudprovider
+    namespace: shoot--foo--bar
+  providerConfig:
+    apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
+    kind: InfrastructureConfig
+    resourceGroup:
+      name: mygroup
+    networks:
+      vnet: # specify either 'name' or 'cidr'
+      # name: my-vnet
+        cidr: 10.250.0.0/16
+      workers: 10.250.0.0/19
+status:
+  state:
+    terraformState: |
```

Comment:
The data structure in the `state` field in each custom resource should depend on the respective extension controller.

For the Infrastructure controller currently there is no type other than `terraform` so we could just use `value`

For the Worker it could be:
```
state:
    - type: machineClass
       value: < RawExtension containing a machineClass object >
    - type: machineDeployment
       value: < RawExtension containing a machineDeployment object 
    - type: machineSet
       value: < RawExtension containing a machineSet object >
    - type: machine
       value: < RawExtension containing a machine object >
    - type: machine
       value: < RawExtension containing a machine object >
``` 
We could replace `type` with `kind` here. 

The alternative to the type/kind approach would be:
```
state:
    machineClasses:
        - < RawExtension containing a machineClass object >
    machineDeployments:
        - < RawExtension containing a machineDeployment object >
    machinesets:
       - < RawExtension containing a machineSet object >
    machines:
      - < RawExtension containing a machine object >
      - < RawExtension containing a machine object >
``` 

If we want to use references then the ShootState controller has to inspect the state field of a CR to be able to decide whether to sync a configmap (or secret) to the garden cluster or to just copy the state data to the `ShootState` resource. This forces the structure of the state to be something like: `.state.value`, `.state.valueRef` (could be done with an array as in the Worker example above)